kv/kvserver: TestReplicaLaziness failure when run under race on Pebble

Failure seen on CI: https://teamcity.cockroachdb.com/viewLog.html?buildId=1922468&buildTypeId=Cockroach_UnitTests

```
--- FAIL: TestReplicaLaziness (0.07s)
    replica_test.go:4497: unexpected error: cannot read undeclared span /Min
        declared:
        read global: /M{in-ax} at 0.000000123,3
        
        stack:
        goroutine 23229 [running]:
        runtime/debug.Stack(0xc000aa2ed0, 0x0, 0x0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0xab
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.(*SpanSet).checkAllowed(0xc000aa2ea0, 0x0, 0x9b1b960, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc003421480, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/spanset.go:293 +0x3ad
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.(*SpanSet).CheckAllowedAt(0xc000aa2ea0, 0x0, 0x9b1b960, 0x0, 0x0, 0x0, 0x0, 0x0, 0x7b, 0x3, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/spanset.go:235 +0x116
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset.(*Iterator).SeekGE(0xc00070b180, 0x9b1b960, 0x0, 0x0, 0x0, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/spanset/batch.go:74 +0x367
        github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).iterSeek(0xc003eda580, 0x9b1b960, 0x0, 0x0, 0x0, 0x0, 0xbdaf2c)
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:637 +0xb4
        github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).scan(0xc003eda580, 0x2a0, 0xc003eda580, 0x7f8df553dbb8)
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:176 +0x7df
        github.com/cockroachdb/cockroach/pkg/storage.mvccScanToBytes(0x72928c0, 0xc003980330, 0x7329160, 0xc00070b180, 0x9b1b960, 0x0, 0x0, 0x9b1b9a2, 0x2, 0x2, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:2366 +0x48a
        github.com/cockroachdb/cockroach/pkg/storage.mvccScanToKvs(0x72928c0, 0xc003980330, 0x7329160, 0xc00070b180, 0x9b1b960, 0x0, 0x0, 0x9b1b9a2, 0x2, 0x2, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:2396 +0x13d
        github.com/cockroachdb/cockroach/pkg/storage.MVCCScan(0x72928c0, 0xc003980330, 0x72ddda0, 0xc000aa2fc0, 0x9b1b960, 0x0, 0x0, 0x9b1b9a2, 0x2, 0x2, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:2539 +0x276
        github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval.Scan(0x72928c0, 0xc003980330, 0x72ddda0, 0xc000aa2fc0, 0x733af40, 0xc00059afc0, 0x7b, 0x3, 0x0, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval/cmd_scan.go:60 +0x984
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateCommand(0x72928c0, 0xc003980330, 0x0, 0x0, 0x0, 0x73277c0, 0xc000aa2fc0, 0x733af40, 0xc00059afc0, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_evaluate.go:473 +0x74c
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateBatch(0x72928c0, 0xc003980330, 0x0, 0x0, 0x73277c0, 0xc000aa2fc0, 0x733af40, 0xc00059afc0, 0x0, 0xc001b83280, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_evaluate.go:241 +0x492
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatchWithServersideRefreshes(0xc002f10000, 0x72928c0, 0xc003980330, 0x73277c0, 0xc000aa2fc0, 0x733af40, 0xc00059afc0, 0xc001b83280, 0xc000aa2ea0, 0xc001754178, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_read.go:129 +0x148
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeReadOnlyBatch(0xc002f10000, 0x72928c0, 0xc003980330, 0xc001b83280, 0x0, 0x0, 0xc0010e9b40, 0x100000001, 0x1, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_read.go:66 +0x597
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc002f10000, 0x72928c0, 0xc003980330, 0xc001b83280, 0x6358e98, 0x0, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:247 +0x5b7
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc002f10000, 0x72928c0, 0xc003e00390, 0x1, 0xc001b83280, 0x0, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:98 +0x675
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(0xc002f10000, 0x7292840, 0xc000074130, 0x7b, 0x3, 0x0, 0x0, 0x0, 0x1, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:36 +0xce
        github.com/cockroachdb/cockroach/pkg/kv.Wrap.func1(0x7292840, 0xc000074130, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:426 +0x109
        github.com/cockroachdb/cockroach/pkg/kv.SenderFunc.Send(0xc000258900, 0x7292840, 0xc000074130, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/sender.go:362 +0xac
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.TestReplicaLaziness.func1(0x6359b18)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_test.go:4496 +0x2f4
        github.com/cockroachdb/cockroach/pkg/kv/kvserver.TestReplicaLaziness(0xc003050600)
        	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_test.go:4515 +0xd0
        testing.tRunner(0xc003050600, 0x6359b20)
        	/usr/local/go/src/testing/testing.go:909 +0x19a
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:960 +0x652
```

I suspect the reason we're not seeing this under RocksDB is because we don't go through the spanset iterator for the RocksDB `MVCCScan`. I'm presuming this is merely a problem with the test and not something deeper.

I haven't tested, but believe this can be reproduced via `COCKROACH_STORAGE_ENGINE=pebble make stressrace PKG=./pkg/kv/kvserver TESTS=TestReplicaLaziness`.

@tbg Assigning to you for triage and guidance. If this is easy to fix, I'm happy to do so. Just tell me what to do.

--- a/pkg/kv/kvserver/replica_test.go
+++ b/pkg/kv/kvserver/replica_test.go
@@ -4513,7 +4513,7 @@ func TestReplicaLaziness(t *testing.T) {
 	})
 
 	testWithAction(func() roachpb.Request {
-		scan := scanArgs(roachpb.KeyMin, roachpb.KeyMax)
+		scan := scanArgs(roachpb.Key("a"), roachpb.Key("b"))
 		return scan
 	})
 }

--- a/pkg/kv/kvserver/spanset/batch.go
+++ b/pkg/kv/kvserver/spanset/batch.go
@@ -102,7 +102,7 @@ func (i *Iterator) Valid() (bool, error) {
 	}
 	ok, err := i.i.Valid()
 	if err != nil {
-		return false, i.err
+		return false, err
 	}
 	return ok && !i.invalid, nil
 }
@@ -110,13 +110,15 @@ func (i *Iterator) Valid() (bool, error) {
 // Next is part of the engine.Iterator interface.
 func (i *Iterator) Next() {
 	i.i.Next()
-	if i.spansOnly {
-		if i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}) != nil {
-			i.invalid = true
-		}
-	} else {
-		if i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}, i.ts) != nil {
-			i.invalid = true
+	if ok, _ := i.i.Valid(); ok {
+		if i.spansOnly {
+			if i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}) != nil {
+				i.invalid = true
+			}
+		} else {
+			if i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}, i.ts) != nil {
+				i.invalid = true
+			}
 		}
 	}
 }
@@ -124,13 +126,15 @@ func (i *Iterator) Next() {
 // Prev is part of the engine.Iterator interface.
 func (i *Iterator) Prev() {
 	i.i.Prev()
-	if i.spansOnly {
-		if i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}) != nil {
-			i.invalid = true
-		}
-	} else {
-		if i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}, i.ts) != nil {
-			i.invalid = true
+	if ok, _ := i.i.Valid(); ok {
+		if i.spansOnly {
+			if i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}) != nil {
+				i.invalid = true
+			}
+		} else {
+			if i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}, i.ts) != nil {
+				i.invalid = true
+			}
 		}
 	}
 }
@@ -138,13 +142,15 @@ func (i *Iterator) Prev() {
 // NextKey is part of the engine.Iterator interface.
 func (i *Iterator) NextKey() {
 	i.i.NextKey()
-	if i.spansOnly {
-		if i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}) != nil {
-			i.invalid = true
-		}
-	} else {
-		if i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}, i.ts) != nil {
-			i.invalid = true
+	if ok, _ := i.i.Valid(); ok {
+		if i.spansOnly {
+			if i.spans.CheckAllowed(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}) != nil {
+				i.invalid = true
+			}
+		} else {
+			if i.spans.CheckAllowedAt(SpanReadOnly, roachpb.Span{Key: i.UnsafeKey().Key}, i.ts) != nil {
+				i.invalid = true
+			}
 		}
 	}
 }

--- a/pkg/storage/pebble_mvcc_scanner.go
+++ b/pkg/storage/pebble_mvcc_scanner.go
@@ -16,7 +16,6 @@ import (
 	"sort"
 	"sync"
 
-	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
@@ -525,7 +524,7 @@ func (p *pebbleMVCCScanner) advanceKeyAtEnd() bool {
 		// Iterating to the next key might have caused the iterator to reach the
 		// end of the key space. If that happens, back up to the very last key.
 		p.peeked = false
-		p.parent.SeekLT(MVCCKey{Key: keys.MaxKey})
+		p.parent.SeekLT(MVCCKey{Key: p.end})
 		if !p.updateCurrent() {
 			return false
 		}
@@ -665,7 +664,7 @@ func (p *pebbleMVCCScanner) iterNext() bool {
 		if !p.iterValid() {
 			// We were peeked off the beginning of iteration. Seek to the first
 			// entry, and then advance one step.
-			p.parent.SeekGE(MVCCKey{})
+			p.parent.SeekGE(MVCCKey{Key: p.start})
 			if !p.iterValid() {
 				return false
 			}

